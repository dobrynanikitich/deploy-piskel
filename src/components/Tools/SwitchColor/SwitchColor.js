import React, { Component } from 'react';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classes from './SwitchColor.module.scss';

class SwitchColor extends Component {
    firstColor = React.createRef();
    secondColor= React.createRef();
    componentDidMount() {
        this.firstColor.current.style.backgroundColor = '#000000';
        this.secondColor.current.style.backgroundColor = '#ffffff';
    }

    rgbToHex = (r, g, b) => {
        return `#${((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1)}`;
      }

    changeCurrentColor = (e) => {
        this.props.changeState('currentColor', e.target.value);
        e.target.parentNode.style.backgroundColor = e.target.value;
    }

    changeSecondColor = (e) => {
        e.target.parentNode.style.backgroundColor = e.target.value;
    }

    switchColors = () => {
        const mainColor = this.firstColor.current;
        const secondColor = this.secondColor.current;
        const colorFirst = mainColor.style.backgroundColor;
        const colorSecond = secondColor.style.backgroundColor;
        const colorInRgb = colorSecond.slice(4, -1).split(',');
        const colorInHex = this.rgbToHex(Number(colorInRgb[0]), Number(colorInRgb[1]), Number(colorInRgb[2]));
        mainColor.style.backgroundColor = colorInHex;
        secondColor.style.backgroundColor = colorFirst;

        this.props.changeState('currentColor', mainColor.style.backgroundColor);
    }

    render() {
        return (
            <>
                <div className={classes.SwitchColorBlock}>
                    <label id='firstColor' className={classes.CurrentColor} ref={this.firstColor}>
                        <input type='color' value='#808080' id='mainColor' onChange={(e) => this.changeCurrentColor(e)}/>
                    </label>
                    <label id='secondColor' className={classes.PrevColor} ref={this.secondColor}>
                        <input type='color' value='#808080' id='secondColor' onChange={(e) => this.changeCurrentColor(e)}/>
                    </label>
                </div>
                <div className={classes.SwitchColorsBtn} data-tip='Press to switch colors'>
                    <FontAwesomeIcon icon='exchange-alt' size="2x" onClick={this.switchColors}/>
                </div>
            </>
        );
    }
}

export default SwitchColor;

