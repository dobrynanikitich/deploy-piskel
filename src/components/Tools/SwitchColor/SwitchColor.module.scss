.SwitchColorBlock {
    width: 100%;
    max-width: 135px;
    display: flex;
    justify-content: space-around;
    label {
        width: 60px;
        height: 60px;
        border: 5px solid darkslategray;
        box-sizing: border-box;
    }
    input {
        opacity: 0;
        width: 100%;
        height: 100%;
        box-sizing: border-box;
    }
}

.CurrentColor {
    background-color: #000000;
}

.SwitchColorsBtn {
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    color: white;
}