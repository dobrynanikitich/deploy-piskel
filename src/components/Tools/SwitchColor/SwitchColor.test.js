import React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createSerializer } from 'enzyme-to-json';
import sinon from 'sinon';

import SwitchColor from './SwitchColor';

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

Enzyme.configure({ adapter: new Adapter() });

it('renders correctly', () => {
    const wrapper = mount(
      <SwitchColor 
        changeState={() => {}}
        currentColor={'#000000'}
      />
    );

    expect(wrapper).toMatchSnapshot();
});



it('should return correct hex color', () => {
    const wrapper = mount(
      <SwitchColor 
        changeState={() => {}}
        currentColor={'#000000'}
      />
    );

    expect(wrapper.instance().rgbToHex(0, 0, 0)).toEqual('#000000');
});

it('should execute changestate with correct data', () => {
    const spy = sinon.spy();
    const wrapper = mount(
        <SwitchColor 
            changeState={spy}
            currentColor={'#000000'}
        />
    );

    wrapper.instance().changeCurrentColor({ target: { value: '#000000', parentNode: { style: { backgroundColor: '#000000' } }} });
    expect(spy.calledOnce).toBe(true);
});

it('should change color of the object correctly', () => {
    const obj = {
        target: {
            parentNode: {
                style: {
                    backgroundColor: '#000000',
                }
            },
            value: '#ffffff',
        }
    }
    const wrapper = mount(
        <SwitchColor 
            changeState={() => {}}
            currentColor={'#000000'}
        />
    );

    wrapper.instance().changeSecondColor(obj);
    expect(obj.target.parentNode.style.backgroundColor).toEqual('#ffffff');
});

it('should execute changestate with correct data', () => {
    const spy = sinon.spy();
    const wrapper = mount(
        <SwitchColor 
            changeState={spy}
            currentColor={'#000000'}
        />
    );

    wrapper.instance().switchColors();
    expect(spy.calledOnce).toBe(true);
});