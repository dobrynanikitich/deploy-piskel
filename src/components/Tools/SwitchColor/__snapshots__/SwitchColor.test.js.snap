// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="SwitchColorBlock"
  >
    <label
      className="CurrentColor"
      id="firstColor"
    >
      <input
        id="mainColor"
        onChange={[Function]}
        type="color"
        value="#808080"
      />
    </label>
    <label
      className="PrevColor"
      id="secondColor"
    >
      <input
        id="secondColor"
        onChange={[Function]}
        type="color"
        value="#808080"
      />
    </label>
  </div>,
  <div
    className="SwitchColorsBtn"
    data-tip="Press to switch colors"
  />,
]
`;
